### PR TITLE
fix(chunk_upload): Reduce the chance for races during assembly

### PR DESCRIPTION
**Pull request by jan-auer**
Fixes [SENTRY-6D5](https://sentry.io/sentry/sentry/issues/528067078/)

### CODE REVIEW

File: src/sentry/models/dsymfile.py [hunk 1]
```diff
@@ -297,30 +297,34 @@ def create_dsym_from_id(project, dsym_type, cpu_name, debug_id,
...
+            'project': project
+        }
+
+        if rv is None:
+            try:
+                with transaction.atomic():
+                    rv = ProjectDSymFile.objects.create(**kwargs)
+            except IntegrityError:
+                file.delete()
+                rv = ProjectDSymFile.objects.get(debug_id=debug_id, project=project)
```
Comment (mitsuhiko) [inline src/sentry/models/dsymfile.py]: This means we keep the old file now rather than the new one.

Comment (jan-auer) [reply src/sentry/models/dsymfile.py]: It‘s the same as before.

Comment (mitsuhiko) [reply src/sentry/models/dsymfile.py]: How so? Before we deleted the old file, now we delete the new one no?

Comment (jan-auer) [reply src/sentry/models/dsymfile.py]: I might be missing something obvious here, but this is the relevant section in the old code: https://github.com/getsentry/sentry/blob/93b0e5b4332c241f25af4c63062090001c35f08e/src/sentry/models/dsymfile.py#L322-L323

This branch handles if two concurrent jobs try to add a file at the same time and just lets the faster job win.

**Comment by jan-auer** [comment]
@mitsuhiko Updated based on offline discussion.

**Review by mitsuhiko — APPROVED** [review]

**Comment by jan-auer** [comment]
Either this didn't work, or we have another race condition, since this is still happening:
https://sentry.io/sentry/sentry/issues/528067078/

**Comment by mitsuhiko** [comment]
I will have another look tomorrow. This is odd.

**Comment by jan-auer** [comment]
Had another look at the issue. The problem in the assemble task was clearly resolved: https://sentry.io/sentry/sentry/issues/529591799/. I now suspect that there are just broken dsyms in the database which we might have to clear out.

**Pull request merged**